File: 0698-partition-to-k-equal-sum-subsets/0698-partition-to-k-equal-sum-subsets.cs
// using backtracking with DP
public class Solution {
    int visited = 0;
    HashSet<int>[] seen;  // track if this bitmask has been seen or not (for given k)

    public bool CanPartitionKSubsets(int[] nums, int k) {
        if(nums.Length < k) return false;

        var totalSum = nums.Sum();
        if(totalSum % k != 0) return false;  // it can never be divided in 4

        // sort array in descending
        Array.Sort(nums);
        Array.Reverse(nums);

        var sideSum = totalSum / k;              // allowed sum of each side
        if(nums[0] > sideSum) return false;  // one stick cannot exceed sideSum

        // try to split the nums into k groups
        seen = new HashSet<int>[nums.Length+1];
        for(var i=0; i<=nums.Length; i++) seen[i] = new HashSet<int>();
        return CanPartitionKSubsets(nums, k, sideSum, 0, 0);
    }

    private bool CanPartitionKSubsets(int[] nums, int k, int targetSum, int currSum, int idx){
        // if we have already seen this mask, then useless
        if(seen[k].Contains(visited)) return false;
        seen[k].Add(visited);

        // all pending nums can be put in one side, if we need only one side
        if(k == 1) return true; 
        
        // if current side is done, then look for other sides
        if(currSum == targetSum) return CanPartitionKSubsets(nums, k-1, targetSum, 0, 0);

        // try to fit every remaining matchstick one by one to match targetSum
        for(var i=idx; i<nums.Length; i++){
            var mask = 1 << i;
            if((visited & mask) == mask || (currSum + nums[i]) > targetSum) continue;
            visited |= mask;
            if(CanPartitionKSubsets(nums, k, targetSum, currSum + nums[i], idx+1)) return true;
            visited ^= mask;
        }

        // not possible to make square with current setup
        return false;
    }
}

File: 0698-partition-to-k-equal-sum-subsets/0698-partition-to-k-equal-sum-subsets.py
class Solution:
    def canPartitionKSubsets(self, nums: List[int], k: int) -> bool:
        if len(nums) < k: return False

        totalSum = sum(nums)
        if totalSum % k != 0: return False  # it can never be divided in 4

        # sort array in descending
        nums.sort(reverse=True)

        sideSum = totalSum / k              # allowed sum of each side
        if nums[0] > sideSum: return False  # one stick cannot exceed sideSum

        # try to split the nums into k groups
        self.seen = [set() for _ in range(len(nums) + 1)]
        return self.dfs(nums, k, sideSum, 0, 0, 0)

    def dfs(self, nums: List[int], k: int, targetSum: int, visited: int, currSum: int, idx: int):
        # if we have already seen this mask, then useless
        if visited in self.seen[k]: return False
        self.seen[k].add(visited)

        # all pending nums can be put in one side, if we need only one side
        if k == 1: return True 
        
        # if current side is done, then look for other sides
        if currSum == targetSum: return self.dfs(nums, k-1, targetSum, visited, 0, 0)

        # try to fit every remaining matchstick one by one to match targetSum
        for i in range(len(nums)):
            mask = 1 << i
            if (visited & mask) == mask or (currSum + nums[i]) > targetSum: continue
            visited |= mask
            if self.dfs(nums, k, targetSum, visited, currSum + nums[i], idx+1): return True
            visited ^= mask

        # not possible to make square with current setup
        return False